modules/git/repo_archive.go

```diff
@@ -49,17 +54,86 @@ func ToArchiveType(s string) ArchiveType {
 	return 0
 }
 
-// CreateArchive create archive content to the target path
-func (repo *Repository) CreateArchive(ctx context.Context, format ArchiveType, target io.Writer, usePrefix bool, commitID string) error {
-	if format.String() == "unknown" {
-		return fmt.Errorf("unknown format: %v", format)
+// annexedFiles returns all files in the annex which are present in gitea.
+func (repo *Repository) annexedFiles(ctx context.Context, commitID string) ([]string, error) {
+	// annex_cmd := NewCommand(ctx, "annex", "find", "--anything", "--print0")
+	cmd := NewCommand(ctx, "annex", "find", "--print0")
+	cmd.AddOptionFormat("--branch=%s", commitID)
+
+	var stderr strings.Builder
+	var stdout strings.Builder
+	err := cmd.Run(&RunOpts{
+		Dir:    repo.Path,
+		Stdout: &stdout,
+		Stderr: &stderr,
+	})
+	if err != nil {
+		return nil, ConcatenateError(err, stderr.String())
+	}
+	annexedFiles := strings.Split(stdout.String(), "\x00")
+	annexedFiles = annexedFiles[:len(annexedFiles)-1]
+
+	return annexedFiles, nil
+}
+
+func isIn(l []string, s string) bool {
+	for _, e := range l {
+		if e == s {
+			return true
+		}
+	}
+	return false
+}
+
+// getAnnexKeyForFile returns the annex key for a file in the repo at commitID.
+func (repo *Repository) getAnnexKeyForFile(ctx context.Context, commitID string, file string) (string, error) {
+	// Get the key of the annexed file.
+	// I would prefer to use git annex lookupkey here, but that does not work on bare repositories.
+	// TODO: properly handle git annex pointer files, this works for basic ones though
+	cmd := NewCommand(ctx, "show")
+	cmd.AddDynamicArguments(fmt.Sprintf("%s:%s", commitID, file))
+	var stderr strings.Builder
+	var stdout strings.Builder
+	err := cmd.Run(&RunOpts{
+		Dir:    repo.Path,
+		Stdout: &stdout,
+		Stderr: &stderr,
+	})
+	if err != nil {
+		return "", ConcatenateError(err, stderr.String())
+	}
+	tmp := strings.Split(stdout.String(), "/")
+	annex_key := strings.TrimSuffix(tmp[len(tmp)-1], "\n")
+	return annex_key, nil
+}
```

Comment:
I wrote this same function in

https://github.com/neuropoly/gitea/blob/4d3bf2dce5bdeea719c8dede2a4a0fd41e88fc77/modules/annex/annex.go#L48

I also noticed that `lookupkey` was no good for bare repos, sadly. I guess I didn't leave a comment about it though!

I used `NewCommandContextNoGlobals` in mine because by default gitea blocks git-credential(1)

https://github.com/neuropoly/gitea/blob/4d3bf2dce5bdeea719c8dede2a4a0fd41e88fc77/modules/git/git.go#L187C2-L190

which is fair, except that I need it to get git-annex to talk to private HTTP remotes.